Improve :has() pseudo-class documentation (#20889)
mdn/content: files/en-us/web/css/selector_list/index.md

{{CSSRef}}

> The CSS **selector list** (`,`) selects all the matching nodes. A selector list is a comma-separated list of selectors. 

Reviewer: ```suggestion
The CSS **selector list** (`,`) selects all the matching nodes. A selector list is a comma-separated list of selectors.
```

```css
element, element, element { style properties }

:functional-pseudoclass(element, element, element) { style properties }
```

## Description

When more than one selectors shares the same declarations, they can be grouped together into a comma-separated list. Selector lists can also be passed as parameters to some functional CSS pseudo-classes. White space may appear before and/or after the comma.

The following three declarations are equivalent:

```css
span {
  border: red 2px solid;
}
div {
  border: red 2px solid;
}
```

```css
span,
div {
  border: red 2px solid;
}
```

```css
:is(span, div) {
  border: red 2px solid;
}
```

## Examples

When applying the same styles to elements matching different criteria, grouping the selectors in a comma-separated list can improve consistency while reducing the size of style sheets.

### Single line grouping

This example shows grouping selectors in a single line using a comma-separated list.

```css
h1, h2, h3, h4, h5, h6 {
  font-family: helvetica;
}
```

### Multi line grouping

This example shows grouping selectors in multiple lines using a comma-separated list.

```css
#main,
.content,
article,
h1 + p {
  font-size: 1.1em;
}
```

> ## Valid and invalid selector lists

Reviewer: ```suggestion
## Invalid and forgiving selector lists
```

}
```

## Syntax

```css
element, element, element { style properties }
```

## Description

> To reduce the size of style sheets, one can group selectors in a comma-separated list.

Reviewer: ```suggestion
When applying styles to elements matching different criteria, group the selectors in a comma-separated list. This improves consistency while reducing the size of style sheets.
```

h3 { font-family: sans-serif; }
```

```css
h1, h2:maybe-unsupported, h3 { font-family: sans-serif; }
```

They are not equivalent. In the first rule set, styles will be applied on the `h1` and `h3` elements, but the `h2` rule will not be parsed. However, in the second rule set, because one selector in the list is invalid, the entire rule will not be parsed and  no style will be applied to the `h1` and `h3` elements as well.

### Forgiving selector list

A way to remedy the [invalid selector list](#invalid_selector_list) problem is to use the {{CSSxRef(":is", ":is()")}} or the {{CSSxRef(":where", ":where()")}} pseudo-class, which accept a forgiving selector list. Each selector in a forgiving selector list is parsed individually. So any invalid selectors in the list are ignored and the ones that are valid are used.

Carrying on from the previous example, the following two CSS rule sets are now equivalent:

```css
h1 { font-family: sans-serif; }
h2:maybe-unsupported { font-family: sans-serif; }
h3 { font-family: sans-serif; }
```

```css
:is(h1, h2:maybe-unsupported, h3) { font-family: sans-serif; }
```

### Forgiving relative selector list

> A forgiving relative selector list is similar to the [forgiving selector list](#forgiving_selector_list); it just parses the list components as [relative selectors](/en-US/docs/Web/CSS/CSS_Selectors#relative_selector) rather than [complex selectors](/en-US/docs/Web/CSS/CSS_Selectors#complex_selector).

Reviewer: ```suggestion
A forgiving relative selector list is similar to the [forgiving selector list](#forgiving_selector_list); it just parses the list components as [relative selectors](/en-US/docs/Web/CSS/CSS_Selectors#relative_selector), which begin with an explicit or implied combinator, rather than [complex selectors](/en-US/docs/Web/CSS/CSS_Selectors#complex_selector).
```

```

```css
h1, h2:maybe-unsupported, h3 {
  font-family: sans-serif;
}
```

> They are not equivalent. In the first rule set, styles will be applied on the `h1` and `h3` elements, but the `h2` rule will not be parsed. However, in the second rule set, because one selector in the list is invalid, the entire rule will not be parsed and no style will be applied to the `h1` and `h3` elements as well.

Reviewer: ```suggestion
They are not equivalent. In the first rule set, styles will be applied on the `h1` and `h3` elements, but the `h2` rule will not be parsed. However, in the second rule set, because one selector in the list is invalid, the entire rule will not be parsed and no style will be applied to the `h1` and `h3` elements as well; the entire style block is ignored.
```

}
```

> ### Forgiving relative selector list

Reviewer: ```suggestion
The difference between the two is that the specificity of `:is()` is its most specific argument whereas the `:where()` selector and the forging selector list parameter do not add any specificity weight. 

### Forgiving relative selector list
```

A forgiving relative selector list is similar to the [forgiving selector list](#forgiving_selector_list); it just parses the list components as [relative selectors](/en-US/docs/Web/CSS/CSS_Selectors#relative_selector), which begin with an explicit or implied combinator, rather than [complex selectors](/en-US/docs/Web/CSS/CSS_Selectors#complex_selector).

## Examples

> 

Reviewer: what is the purpose of these two examples?

Reviewer: These should come first, not last, as they are used in all the other examples and explanations. Also, they are part of the definition, so not in an examples section, but rather included under their own heading with a brief sentence, and then a statement saying either is ok, with some company guidelines and linters preferring one over the other, so basically pick one, and stick with it.

```

```css
h1,

> h2:maybe-unsupported,

Reviewer: changed from unsupported to invalid because some selectors are not supported but are not invalid, such as FF neither supporting nor failing on `-webkit` prefixed pseudo-elements. So, while neight `::foobar` nor `::-webkit-foobar` is supported in FF, FF only considers the former invalid and will ignore the whole declaration.

## See also

- The {{CSSxRef(":is", ":is()")}} and {{CSSxRef(":where", ":where()")}} pseudo-classes accept a forgiving selector list.

> - [Forgiving selector list](https://drafts.csswg.org/selectors/#typedef-forgiving-selector-list)

Reviewer: Link the forgiving selector list from the definition to the spec if you want, but don't add it in the see also as it doesn't provide any more detail than the MDN description. 
```suggestion
```
